File: internal/execution/adapters/grok_effort.go
package adapters

import (
	"fmt"
	"os"
	"strings"

	"github.com/nightgauge/nightgauge/internal/models"
)

// Grok CLI extra effort rungs that are not Nightgauge EFFORT_LEVELS.
// They collapse to "low" for registry / thinking-interlock purposes (#523).
func mapGrokEffortToNightgauge(effort string) string {
	e := strings.ToLower(strings.TrimSpace(effort))
	switch e {
	case "":
		return ""
	case "none", "minimal":
		return "low"
	case "low", "medium", "high", "xhigh", "max":
		return e
	default:
		return ""
	}
}

// grokCliEffortFlag is the value to pass as `grok --effort`. SYNTAX validation
// only — membership in the Grok CLI's documented vocabulary. Whether the
// resolved model actually serves a rung is the registry's call, enforced by
// validateGrokEffort before spawn (#569); this list must never grow back into
// a second authority on that question.
func grokCliEffortFlag(effort string) string {
	e := strings.ToLower(strings.TrimSpace(effort))
	switch e {
	case "none", "minimal", "low", "medium", "high", "xhigh", "max":
		return e
	default:
		return ""
	}
}

// validateGrokEffort enforces the registry's supported_efforts for the model a
// Grok dispatch will actually serve (#569). The provider-global
// NIGHTGAUGE_GROK_EFFORT used to pass the static vocabulary filter above and
// reach `grok --effort` unchecked — a rung the resolved model does not declare
// (e.g. `xhigh` against grok-4.5, which tops out at `high`) died as #532's
// signature: exit 1 in seconds, no work, nothing classified. The registry knew
// and was never asked.
//
// Rules, in order:
//   - No explicit effort → nothing to enforce.
//   - A value outside the Grok CLI vocabulary is dropped by BuildCommand's
//     syntax filter exactly as before, but the drop is logged, never silent.
//   - The model is resolved the same way BuildCommand resolves it (band → the
//     registry's xai model, concrete id → itself). A model with NO registry
//     descriptor passes through with a logged warning, never a hard failure
//     (#336) — there is nothing to validate against.
//   - Enforcement runs AFTER normalization (#523): grok-native none/minimal
//     collapse to "low" before the rung is compared to the ladder.
//   - `supported_efforts: []` is a positive declaration — no effort axis — so
//     ANY explicit effort is rejected (#336).
//   - A normalized rung missing from the declared ladder fails closed with an
//     error naming the model, the requested effort, and the ladder — never a
//     silent pass-through, never a silent downgrade (#75).
func validateGrokEffort(model, effort string) error {
	e := strings.ToLower(strings.TrimSpace(effort))
	if e == "" {
		return nil
	}
	if grokCliEffortFlag(e) == "" {
		fmt.Fprintf(os.Stderr,
			"[grok] warning: NIGHTGAUGE_GROK_EFFORT=%q is not a Grok CLI effort (none|minimal|low|medium|high|xhigh|max) — the --effort flag will be omitted and the provider default used\n",
			effort)
		return nil
	}
	resolved, ok := models.Resolve(models.ProviderForAdapter("grok"), strings.TrimSpace(model))
	if !ok {
		fmt.Fprintf(os.Stderr,
			"[grok] warning: model %q has no registry descriptor — cannot verify effort %q against supported_efforts; passing through\n",
			model, effort)
		return nil
	}
	if len(resolved.SupportedEfforts) == 0 {
		return fmt.Errorf(
			"effort %q is not supported by model %q: the model declares no effort axis (supported_efforts: []); unset NIGHTGAUGE_GROK_EFFORT or route to a model with an effort ladder",
			e, resolved.ID,
		)
	}
	normalized := mapGrokEffortToNightgauge(e)
	for _, s := range resolved.SupportedEfforts {
		if s == normalized {
			return nil
		}
	}
	note := ""
	if normalized != e {
		note = fmt.Sprintf(" (normalized to %q)", normalized)
	}
	return fmt.Errorf(
		"effort %q%s is not supported by model %q (supports: %s); choose a supported level or route to a model that accepts %q",
		e, note, resolved.ID, strings.Join(resolved.SupportedEfforts, ", "), e,
	)
}

File: packages/nightgauge-sdk/src/cli/adapters/grokEffort.ts
/**
 * Map Nightgauge effort vocabulary onto Grok Build CLI `--effort` values, and
 * enforce the model registry's `supported_efforts` before dispatch.
 *
 * Nightgauge {@link EFFORT_LEVELS} are `low|medium|high|xhigh|max`.
 * Grok also accepts `none` and `minimal`. Those extra rungs collapse to
 * `low` for registry / interlock purposes; spawn still forwards the
 * vendor flag when the caller already used Grok vocabulary.
 *
 * The static {@link GROK_CLI_EFFORTS} list is SYNTAX validation only — the
 * vocabulary the CLI documents for `--effort`. Whether the resolved model
 * actually serves a rung is the registry's call (#569): a rung the model does
 * not declare used to sail through this filter and die inside the CLI as
 * #532's signature (`unknown effort level 'xhigh'`, exit 1 in seconds, no
 * work, nothing classified). {@link grokCliEffortFlag} now consults the
 * registry and fails closed BEFORE spawn instead.
 *
 * @see Issue #523 - grok-native rungs collapse to `low`
 * @see Issue #569 - supported_efforts enforced at every dispatch path
 */

import { EFFORT_LEVELS } from "../../eval/modelEvalSchemas.js";
import { getModelDescriptor } from "../../eval/modelRegistry.js";
import { AdapterError } from "./errors.js";

/** Effort flags the Grok CLI documents for `--effort` / `--reasoning-effort`. */
export const GROK_CLI_EFFORTS = [
  "none",
  "minimal",
  "low",
  "medium",
  "high",
  "xhigh",
  "max",
] as const;
export type GrokCliEffort = (typeof GROK_CLI_EFFORTS)[number];

const GROK_DOCS_URL = "https://docs.x.ai/build/overview";

/**
 * Collapse a Grok or Nightgauge effort string onto Nightgauge's vocabulary.
 * `none` and `minimal` become `low`. Unknown values return undefined.
 */
export function mapGrokEffortToNightgauge(effort: string | undefined): string | undefined {
  if (!effort) return undefined;
  const e = effort.trim().toLowerCase();
  if (e === "none" || e === "minimal") return "low";
  if ((EFFORT_LEVELS as readonly string[]).includes(e)) return e;
  return undefined;
}

/**
 * The `--effort` value to pass the Grok CLI. Nightgauge rungs pass through;
 * `none`/`minimal` stay as Grok-native flags.
 *
 * Enforcement (#569), applied AFTER the #523 normalization — the normalized
 * rung must be a member of the RESOLVED model's `supported_efforts`:
 *
 * - `model` defaults to the same env resolution the adapter dispatches with
 *   (`NIGHTGAUGE_GROK_MODEL` → `NIGHTGAUGE_MODEL`); band names resolve to the
 *   registry's xai model, concrete ids to themselves.
 * - A model with NO registry descriptor passes through with a logged warning,
 *   never a hard failure (#336) — there is nothing to validate against.
 * - `supported_efforts: []` is a positive declaration ("no effort axis"), so
 *   ANY explicit effort throws (#336).
 * - A rung the model does not declare throws an {@link AdapterError}
 *   (`CONFIG_INVALID`) naming the model, the requested effort, and the
 *   declared ladder — never a silent pass-through, never a silent downgrade
 *   (#75).
 * - A value outside the CLI vocabulary is dropped as before (returns
 *   undefined), with a logged warning instead of silence.
 */
export function grokCliEffortFlag(
  effort: string | undefined,
  model: string | undefined = process.env.NIGHTGAUGE_GROK_MODEL ?? process.env.NIGHTGAUGE_MODEL
): GrokCliEffort | undefined {
  if (!effort) return undefined;
  const e = effort.trim().toLowerCase();
  if (!(GROK_CLI_EFFORTS as readonly string[]).includes(e)) {
    if (e) {
      console.warn(
        `[grok-adapter] warning: effort '${e}' is not a Grok CLI effort ` +
          `(${GROK_CLI_EFFORTS.join("|")}) — the --effort flag will be omitted ` +
          `and the provider default used`
      );
    }
    return undefined;
  }

  const trimmedModel = model?.trim();
  const descriptor = trimmedModel ? getModelDescriptor(trimmedModel, "xai") : undefined;
  if (!descriptor) {
    console.warn(
      `[grok-adapter] warning: model '${trimmedModel || "(adapter default)"}' has no ` +
        `registry descriptor — cannot verify effort '${e}' against supported_efforts; ` +
        `passing through`
    );
    return e as GrokCliEffort;
  }

  const ladder = descriptor.supported_efforts;
  if (ladder.length === 0) {
    throw new AdapterError(
      `Effort '${e}' is not supported by model '${descriptor.id}': the model declares ` +
        `no effort axis (supported_efforts: []).\n` +
        `Fix: unset NIGHTGAUGE_GROK_EFFORT or route to a model with an effort ladder.`,
      "CONFIG_INVALID",
      "Grok",
      GROK_DOCS_URL
    );
  }

  const normalized = mapGrokEffortToNightgauge(e);
  if (!normalized || !(ladder as readonly string[]).includes(normalized)) {
    const note = normalized && normalized !== e ? ` (normalized to '${normalized}')` : "";
    throw new AdapterError(
      `Effort '${e}'${note} is not supported by model '${descriptor.id}' ` +
        `(supports: ${ladder.join(", ")}).\n` +
        `Fix: set NIGHTGAUGE_GROK_EFFORT to a supported level or route to a model ` +
        `that accepts '${e}'.`,
      "CONFIG_INVALID",
      "Grok",
      GROK_DOCS_URL
    );
  }

  return e as GrokCliEffort;
}
